pkg/cri/server/restart.go

```diff
@@ -468,8 +468,14 @@ func (c *criService) loadImages(ctx context.Context, cImages []containerd.Image)
 				return
 			}
 			if !unpacked {
-				log.G(ctx).Warnf("The image %s is not unpacked.", i.Name())
```

Comment:
What is the pros/cons of unpacking here? Do we have reasons to keep it disabled by default other than making the behavior backward-compatible?

I feel that exposing the tuning knob is handwavy for containerd users.

Reply:
This was discussed in #6710. 

From @fuweid 
> Looks good to me to have option to enable.
> 
> My concern is that if the unpack takes long time and no response to kubelet, kubelet will report that node is not-ready.

So pros would be giving users an option to automatically unpack the image on recovery, but the cons would be unpacking process may take long enough to break an existing workflow. Making it optional seems balance the two, unless there is a better option.

Reply:
Sorry, I meant, why do users want to unpack this image here? Would it make the recovery process more complete, but slower and/or unsafe?

In addition to that, how does this not-yet-unpacked scenario work today? Would it bubble up to Kubernetes control plane?

Reply:
> Sorry, I meant, why do users want to unpack this image here? Would it make the recovery process more complete, but slower and/or unsafe?

@AkihiroSuda could you comment on this please?

> 
> In addition to that, how does this not-yet-unpacked scenario work today? Would it bubble up to Kubernetes control plane?

AFAIK unpacked images are simply ignored today, except a warning message is logged by containerd.

Reply:
> why do users want to unpack this image here

When an image was created in the k8s.io namespace by a non-CRI client without unpacking it, the CRI plugin can never use the image.
So the CRI should be able to automatically unpack it.

Reply:
after check the code, I think the CreateContainer will unpack the image if not-found.

https://github.com/containerd/containerd/blob/main/pkg/cri/server/container_create.go#L195

```
// WithNewSnapshot wraps `containerd.WithNewSnapshot` so that if creating the                                       
// snapshot fails we make sure the image is actually unpacked and and retry.                                        
func WithNewSnapshot(id string, i containerd.Image, opts ...snapshots.Opt) containerd.NewContainerOpts {                                    
        f := containerd.WithNewSnapshot(id, i, opts...)                                                             
        return func(ctx context.Context, client *containerd.Client, c *containers.Container) error {                                    
                if err := f(ctx, client, c); err != nil {                                                           
                        if !errdefs.IsNotFound(err) {                                                               
                                return err                                                                            
                        }                                                                                           
                                                                                                                    
                        if err := i.Unpack(ctx, c.Snapshotter); err != nil {                                        
                                return fmt.Errorf("error unpacking image: %w", err)                                    
                        }                                                                                           
                        return f(ctx, client, c)                                                                    
                }                                                                                                   
                return nil                                                                                          
        }                                                                                                                             
}
```

We don't know what's the size of image and how long it will take to unpack.
So I would like to unpack the image when use it.

Reply:
Thanks @fuweid. that makes sense, implementing the unpacking logic to container creation time seems much safer, given the concern of the original approach.

Since you already have the code, would you like me to continue this PR by incorporate your patch or you'd like to create a PR by yourself? I am happy either way :) Thanks!

Reply:
hi @henry118 just do it in your patch ~ thanks!

Reply:
Checking the code again, the retry logic mentioned above already exists today. As discussed it's not considered safe to unpack during recovery. I'll withdraw this PR.

Reply:
I'm not sure I understand why it would be unreasonable to have an option for this, and I think generally it would actually be a reasonable thing to turn on, though giving administrators the choice seems reasonable https://github.com/containerd/containerd/issues/6710#issuecomment-1132028575

Non-unpacked images generate a warning, then are unpacked at the last minute when attempting to create the container.

In my use case, this happens because we don't know which snapshotter will be used at import time. 
But after that we would enable auto-unpack (which should make time-to-node-ready _faster_).

We could do something without this change (we can auto-unpack in some external tool), but we'll still be logging this relatively useless warning AND not doing anything about it in containerd when we could, which feels bad.

In what concrete scenario is this unsafe?
What makes it unacceptable as an opt-in as in this current PR?

Reply:
| In my use case, this happens because we don't know which snapshotter will be used at import time.
But after that we would enable auto-unpack (which should make time-to-node-ready faster).

would like to add runtime handler to the image services.. as an option probably open a KEP for it soon so we can have n-snapshotters one per runtime handler.

... the warn could be made a debug message 

| In what concrete scenario is this unsafe?

After seeing this message today "May 17 13:37:08 ip-10-161-14-214.eu-central-1.compute.internal systemd[1]: containerd.service start operation timed out. Terminating." on a new (probably old) issue related to pods stuck in terminating. I'm reminded that some will need containerd to get up fast on a reboot/restart.  

It probably makes sense here to start up a background thread unpacking them (one at a time).